### PR TITLE
hibernate search cannot be used for count query if non-active parameters are present

### LINKED ISSUE #6024: Do not use hibernate search if the search parameter is not an active search parameter
Hibernate search[ ignores search parameters that it can't handle](https://github.com/hapifhir/hapi-fhir/blob/master/hapi-fhir-jpaserver-base/src/main/java/ca/uhn/fhir/jpa/dao/search/ExtendedHSearchSearchBuilder.java#L181).

Because of this, it should not be used for count queries unless it can also be used for the full query.

### PR DESCRIPTION
**Pull request by TipzCM**
closes #6024

### CODE REVIEW

File: hapi-fhir-jpaserver-base/src/main/java/ca/uhn/fhir/jpa/model/search/ExtendedHSearchBuilderConsumeAdvancedQueryClausesParams.java [hunk 1]
```diff
@@ -0,0 +1,47 @@
+package ca.uhn.fhir.jpa.model.search;
+
+import ca.uhn.fhir.jpa.searchparam.SearchParameterMap;
+import ca.uhn.fhir.rest.server.util.ISearchParamRegistry;
+
+public class ExtendedHSearchBuilderConsumeAdvancedQueryClausesParams {
```
Comment (lukedegruchy) [inline hapi-fhir-jpaserver-base/src/main/java/ca/uhn/fhir/jpa/model/search/ExtendedHSearchBuilderConsumeAdvancedQueryClausesParams.java:6]: Please add top level javadoc explaining what this class is for

File: hapi-fhir-jpaserver-base/src/main/java/ca/uhn/fhir/jpa/model/search/ExtendedHSearchBuilderConsumeAdvancedQueryClausesParams.java [hunk 2]
```diff
@@ -0,0 +1,47 @@
+package ca.uhn.fhir.jpa.model.search;
+
+import ca.uhn.fhir.jpa.searchparam.SearchParameterMap;
+import ca.uhn.fhir.rest.server.util.ISearchParamRegistry;
+
```
Comment (lukedegruchy) [inline hapi-fhir-jpaserver-base/src/main/java/ca/uhn/fhir/jpa/model/search/ExtendedHSearchBuilderConsumeAdvancedQueryClausesParams.java:5]: Also, please generate the copyright header from maven and then commit it.

File: hapi-fhir-jpaserver-test-r5/src/test/java/ca/uhn/fhir/jpa/provider/r5/ResourceProviderR5Test.java [hunk 3]
```diff
@@ -207,7 +227,177 @@ public void testErroredSearchReturnsAppropriateResponse() {
...
+			// fullTextSearch means Advanced Hibernate Search
+			myStorageSettings.setFilterParameterEnabled(true);
+			myStorageSettings.setAdvancedHSearchIndexing(true);
+
+			// create custom search parameters - the _filter and _include are needed
+			{
+				IFhirResourceDao<SearchParameter> spDao = myDaoRegistry.getResourceDao("SearchParameter");
+				SearchParameter sp;
+
+				String includeParam = """
```
Comment (lukedegruchy) [inline hapi-fhir-jpaserver-test-r5/src/test/java/ca/uhn/fhir/jpa/provider/r5/ResourceProviderR5Test.java:250]: Nitpick:  can you add @Language("JSON") to these Strings?

**Review by lukedegruchy — APPROVED** [review]

File: hapi-fhir-jpaserver-base/src/main/java/ca/uhn/fhir/jpa/model/search/ExtendedHSearchBuilderConsumeAdvancedQueryClausesParams.java (hunk 1 above)
Comment (TipzCM) [reply hapi-fhir-jpaserver-base/src/main/java/ca/uhn/fhir/jpa/model/search/ExtendedHSearchBuilderConsumeAdvancedQueryClausesParams.java:6]: added

File: hapi-fhir-jpaserver-test-r5/src/test/java/ca/uhn/fhir/jpa/provider/r5/ResourceProviderR5Test.java (hunk 3 above)
Comment (TipzCM) [reply hapi-fhir-jpaserver-test-r5/src/test/java/ca/uhn/fhir/jpa/provider/r5/ResourceProviderR5Test.java:250]: added

File: hapi-fhir-jpaserver-base/src/main/java/ca/uhn/fhir/jpa/model/search/ExtendedHSearchBuilderConsumeAdvancedQueryClausesParams.java (hunk 2 above)
Comment (TipzCM) [reply hapi-fhir-jpaserver-base/src/main/java/ca/uhn/fhir/jpa/model/search/ExtendedHSearchBuilderConsumeAdvancedQueryClausesParams.java:25]: added

File: hapi-fhir-jpaserver-base/src/main/java/ca/uhn/fhir/jpa/dao/search/ExtendedHSearchSearchBuilder.java [hunk 4]
```diff
@@ -58,18 +61,56 @@ public class ExtendedHSearchSearchBuilder {
...
+								.anyMatch(this::isParamTypeSupported);
+			}
+
+			// if not supported, don't use
+			if (!canUseHibernate) {
+				return false;
+			}
+		}
+
+		return canUseHibernate;
```
Comment (tadgh) [inline hapi-fhir-jpaserver-base/src/main/java/ca/uhn/fhir/jpa/dao/search/ExtendedHSearchSearchBuilder.java:113]: issue: this refactor/addition does not maintain the existing behaviour contract. The previous default was to not use HS, and only use it if a known SP could benefit from it. This new iteration defaults to using HS, and _if there are no SPs_ (e.g. in a type-level $everything operation), this method returns true, causing HS to be used for type-level $everything, when it should not be.

**Pull request merged**